src/Scanner.js

```diff
@@ -148,23 +138,57 @@ class DefaultScanner {
         } else {
             const inputs = context.querySelectorAll(FORM_INPUTS_SELECTOR)
             if (inputs.length > this.options.maxInputsPerPage) {
+                this.stopScanner('Too many input fields in the given context, stop scanning', context)
                 return this
             }
             inputs.forEach((input) => this.addInput(input))
         }
         return this
     }
 
+    /**
+     * Stops scanning, switches off the mutation observer and clears all forms
+     * @param {string} reason
+     * @param {...any} rest
+     */
+    stopScanner (reason, ...rest) {
+        this.stopped = true
+
+        if (shouldLog()) {
+            console.log(reason, ...rest)
+        }
+
+        const activeInput = this.device.activeForm?.activeInput
+
+        // remove Dax, listeners, timers, and observers
+        clearTimeout(this.debounceTimer)
+        this.changedElements.clear()
+        this.mutObs.disconnect()
+
+        this.forms.forEach(form => {
+            form.destroy()
+        })
+        this.forms.clear()
+
+        // Bring the user back to the input they were interacting with
+        activeInput?.focus()
+    }
+
     /**
      * @param {HTMLElement|HTMLInputElement|HTMLSelectElement} input
      * @returns {HTMLFormElement|HTMLElement}
      */
     getParentForm (input) {
         if (input instanceof HTMLInputElement || input instanceof HTMLSelectElement) {
-            // Use input.form unless it encloses most of the DOM
-            // In that case we proceed to identify more precise wrappers
-            if (input.form && !isFormLikelyToBeUsedAsPageWrapper(input.form)) {
-                return input.form
+            if (input.form) {
+                // Use input.form unless it encloses most of the DOM
+                // In that case we proceed to identify more precise wrappers
+                if (
+                    this.forms.has(input.form) || // If we've added the form we've already checked that it's not a page wrapper
```

Comment:
By adding this little check we avoid the much more expensive `isFormLikelyToBeUsedAsPageWrapper` check. This significantly speeds up forms with a lot of input fields. For example, further massive gain on your test page with 1000 fields, down to ~70ms on our browser on my machine.